metagraph/src/graph/alignment/dbg_aligner.hpp

```diff
@@ -24,377 +20,153 @@ class IDBGAligner {
-               const SeedGenerator &seed_generator,
-               const std::function<void(DBGAlignment&&)> &callback,
-               const std::function<score_t(const DBGAlignment&)> &get_min_path_score) const;
+    virtual std::unique_ptr<IExtender<node_index>>
+    build_extender(std::string_view query, const ISeeder<node_index> &seeder) const = 0;
 
-    virtual Seeder build_seeder() const = 0;
-    virtual Extender build_extender() const = 0;
+    // Generate seeds, then extend them
+    void align_core(std::string_view query,
+                    const ISeeder<node_index> &seeder,
+                    IExtender<node_index>&& extender,
+                    const LocalAlignmentCallback &callback,
+                    const MinScoreComputer &get_min_path_score) const;
 
     // Align the query sequence in the given orientation (false is forward,
     // true is reverse complement)
-    DBGQueryAlignment align_one_direction(const std::string_view query,
-                                          bool orientation) const;
+    void align_one_direction(DBGQueryAlignment &paths,
+                             bool orientation_to_align,
+                             const ISeeder<node_index> &seeder) const;
+
+    // Align both the forward and reverse complement of the query sequence,
+    // then report the best scoring alignment.
+    void align_best_direction(DBGQueryAlignment &paths,
+                              const ISeeder<node_index> &seeder,
+                              const ISeeder<node_index> &seeder_rc) const;
 
     // Align both forwards and backwards from a given seed. Procedure
     // 1. Given each seed, extend forward to produce an alignment A
     // 2. Reverse complement the alignment to get A', treated like a new seed
     // 3. Extend A' forwards
     // 4. Reverse complement A' to get the final alignment A''
-    DBGQueryAlignment align_both_directions(const std::string_view query) const;
-
-    // Align both the forward and reverse complement of the query sequence,
-    // then report the best scoring alignment.
-    DBGQueryAlignment align_forward_and_reverse_complement(const std::string_view query) const;
+    void align_both_directions(DBGQueryAlignment &paths,
+                               const ISeeder<node_index> &seeder) const;
 
     // Given alignments generated by a generator, add them to a priority queue
     // and add the top ones to paths.
-    virtual void align_aggregate(DBGQueryAlignment &paths,
-                                 const AlignmentGenerator &alignment_generator) const = 0;
-};
+    virtual void
+    align_aggregate(DBGQueryAlignment &paths,
+                    const AlignmentGenerator &alignment_generator) const;
 
+  protected:
+    const DeBruijnGraph &graph_;
+    DBGAlignerConfig config_;
+};
 
 template <class Seeder = ExactSeeder<>,
           class Extender = DefaultColumnExtender<>,
           class AlignmentCompare = std::less<Alignment<>>>
```

Comment:
I would suggest removing this template parameter `AlignmentCompare`.
It's only passed to the priority queue in AlignmentAggregator, and why would you be interested in getting something other than the best alignments?

Reply:
There are multiple parameters which may be used to rank alignments. Score is one of them, but whether you'd prefer long or short alignments is a design choice of the aligner, so I wanted to keep this here as a template parameter.

Reply:
For example, local alignment prefers having shorter, high-scoring alignments, whereas semi-global alignment prefers having the largest part of the query covered.

Reply:
Ok, I was confused by the existence of `std::less<Alignment<>>`, which assumes that it's the alignment you're looking for.

If there are these cases where different rankings may be preferred, it doesn't make sense to implement that operator< in Alignment<>.

Better implement the comparator outside Alignment<> and pass it here `template <..., class AlignmentCompare = BestAlignment>`.
Another way is to pass that comparator directly to Alignment<>.

Do you see what I mean?

Reply:
I added a new struct called `LocalAlignmentLess` to handle this.